ops/testing.py

```diff
@@ -2586,6 +2605,128 @@ class NonAbsolutePathError(Exception):
+        """Write content to a given file path on the remote system.
+
+        See :meth:`ops.model.Container.push` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def list_files(self, path: StrOrPath, *, pattern: Optional[str] = None,
+                   itself: bool = False) -> List['FileInfo']:
+        """Return list of directory entries from given path on remote system.
+
+        See :meth:`ops.model.Container.list_files` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def push_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath) -> None:
+        """Recursively push a local path or files to the remote system.
+
+        See :meth:`ops.model.Container.push_path` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def pull_path(self,
+                  source_path: Union[StrOrPath, Iterable[StrOrPath]],
+                  dest_dir: StrOrPath) -> None:
+        """Recursively pull a remote path or files to the local system.
+
+        See :meth:`ops.model.Container.pull_path` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def exists(self, path: str) -> bool:
+        """Return true if the path exists on the container filesystem.
+
+        See :meth:`ops.model.Container.exists` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def isdir(self, path: str) -> bool:
+        """Return true if a directory exists at the given path on the container filesystem.
+
+        See :meth:`ops.model.Container.isdir` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def make_dir(
+            self, path: str, *, make_parents: bool = False, permissions: Optional[int] = None,
+            user_id: Optional[int] = None, user: Optional[str] = None,
+            group_id: Optional[int] = None, group: Optional[str] = None) -> None:
+        """Create a directory on the remote system with the given attributes.
+
+        See :meth:`ops.model.Container.make_dir` for details.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove_path(self, path: str, *, recursive: bool = False) -> None:
+        """Remove a file or directory on the remote system.
+
+        See :meth:`ops.model.Container.remove_path` for details.
+        """
+        raise NotImplementedError
+
+
+class _ContainerFilesystemProxy:
+    """A proxy providing unrestricted filesystem access for a specified container."""
+
+    def __init__(self, container: model.Container, backend: _TestingPebbleClient):
+        self._container = container
+        self._backend = backend
+
+    def __getattr__(self, name: str):
+        if name not in ContainerFilesystem.__dict__ or name.startswith("_"):
+            raise AttributeError(f'object has no attribute {name!r}')
```

Comment:
The return type of the `get_container_filesystem` is `ops.testing.ContainerFilesystem` instead of  `ops.model.Container`. The `ops.testing.ContainerFilesystem` class just doesn't have other attributes from the definition.

Reply:
Yeah, I think what @weiiwang01 is saying is reasonable. However, I don't think it'd hurt to clarify the message. Perhaps something simple like:

```
raise AttributeError(f'object has no attribute {name!r}; get_container_filesystem only provides access to file-related methods')
```

Reply:
How about changing the error message to `f'{self.__class__.__name__!r} object has no attribute {name!r}'`? This way it will print `'ContainerFilesystem' object has no attribute 'foo'`.

This error message aligns with Python's default message for attribute not found. Moreover, I don't feel like mentioning `get_container_filesystem` in the `ContainerFilesystem` since `ContainerFilesystem` is a dependency of `get_container_filesystem`. So it shouldn't have information about the `get_container_filesystem`.

Reply:
Works for me, thanks!